File: tests/Language/python/comprehensive.py
# Line comment
"""Triple-quoted docstring
multiline
"""
'''Another docstring'''

# Boolean and None literals
is_true = True
is_false = False
is_none = None

# Numbers - decimal, hex, binary, octal, float
dec = 42
hex_num = 0xFF
bin_num = 0b1010
oct_num = 0o755
flt = 3.14
exp = 1.5e10
neg_exp = 2e-5

# Strings
str1 = "Double quoted"
str2 = 'Single quoted'
str3 = """Triple
double"""
str4 = '''Triple
single'''
fstring = f"Value: {value}"
rawstr = r"Raw\nstring"
bytestr = b"Bytes"

# Operators - @ for decorators
@decorator
def func():
    pass

# Two-char operators
a == b
c != d
e <= f
g >= h
i and j
k or l
m // n
o ** p
q += 1
r -= 1
s *= 2
t /= 2
u %= 3
v &= 4
w |= 5
x ^= 6
y <<= 1
z >>= 1
aa //= 2
bb **= 2
cc := val

# Single-char operators
+ - * / % = < > ! & | ^ ~ @ .

# Punctuation
func(arg)
arr[0]
obj.attr
{key: val}
(a, b)
:

# Keywords and control flow
def function(param):
    if condition:
        pass
    elif other:
        pass
    else:
        pass

    for item in items:
        break
    else:
        continue

    while condition:
        pass

    try:
        raise Exception()
    except ValueError as e:
        pass
    finally:
        pass

    with open('file') as f:
        pass

    return result
    yield item

    assert condition
    import module
    from package import name
    global var
    nonlocal outer
    del variable

    class MyClass:
        pass

    async def async_func():
        await coroutine()

    lambda x: x + 1

# String prefix combinations
not_a_fstring = "x#y"  # # should not be comment

File: tests/Language/ruby/comprehensive.rb
# Line comment
=begin
Block comment
multiline
=end

# Boolean and nil
is_true = true
is_false = false
is_nil = nil

# Numbers - decimal, hex, binary, octal, float
dec = 42
hex_num = 0xFF
bin_num = 0b1010
oct_num = 0o755
flt = 3.14
exp = 1.5e10
underscore = 1_000_000

# Symbols
sym = :symbol
quote_sym = :'quoted symbol'

# Strings
str1 = "Double quoted"
str2 = 'Single quoted'
interp = "Value: #{value}"
percent = %{Percent string}
percent_q = %Q{Percent Q}
percent_q_lower = %q{Percent q}

# Heredoc
heredoc = <<HEREDOC
Multi
line
HEREDOC

# Regex
regex = /pattern/
regex_i = /pattern/i
percent_r = %r{pattern}

# Two-char operators
a == b
c != d
e <= f
g >= h
i && j
k || l
m .. n
o ... p
q => r
s += 1
t -= 1
u *= 2
v /= 2
w %= 3
x &= 4
y |= 5
z ^= 6
aa <<= 1
bb >>= 1
cc **= 2
dd <=> ee

# Single-char operators
+ - * / % = < > ! & | ^ ~ @ $ ?

# Punctuation
func(arg)
arr[0]
obj.method
{key: val}
(a, b)
:

# Keywords
def method(param)
  if condition
    puts "yes"
  elsif other
    puts "maybe"
  else
    puts "no"
  end

  unless condition
    puts "not"
  end

  case value
  when 1
    puts "one"
  when 2
    puts "two"
  else
    puts "other"
  end

  for item in items
    break if condition
    next if other
    redo if another
  end

  while condition
    retry
  end

  until condition
    puts "waiting"
  end

  begin
    raise StandardError
  rescue => e
    puts e
  ensure
    cleanup
  end

  return result
  yield block
  super

  class MyClass < Parent
    attr_reader :field
    attr_writer :other
    attr_accessor :both

    def initialize
      @instance = "var"
      @@class_var = "val"
    end

    def self.class_method
      "class"
    end

    private

    def private_method
      "private"
    end

    protected

    def protected_method
      "protected"
    end
  end

  module MyModule
    include OtherModule
    extend AnotherModule
    prepend YetAnother
  end

  alias new_name old_name
  undef method_name
  defined? variable

  lambda { |x| x + 1 }
  -> (x) { x + 1 }

  proc = Proc.new { |x| x }
end
